### PR TITLE
Schema revamp for version 3.0.0

### PR DESCRIPTION
**Pull request by seeker89**
## TL;DR

This is a large revamp for the schema for version 3.0.0, to make powerfulseal more powerful.

This change is not backwards compatible - it tries to fix some hacks which accumulated over the years.

## Details

On the menu:

- [x] merge the node and pod scenarios into a single entity, so that a scenario can affect both nodes and pods
- [x] transform the current `match->filter->action` logic into a type of step
- [x] add a `kubectl` type of action
- [x] add an HTTP probe action
- [x] add basic verification actions in the pod's action phase
- [x] update the implementation to support the new features
  - [x] `kubectl`
  - [x] `checkPodCount`
  - [x] `checkPodState`
  - [x] `probeHTTP`
  - [x] `wait`
- [x] make sure the exit code corresponds to the results of the scenario
- [x] drop `python 3.6`
- [x] cleanup the logs output to be less verbose and more useful
- [x] support `negative` field in property matchers
- [x] update the docs

This takes a slightly different (hopefully more flexible) approach than the solution suggested in https://github.com/bloomberg/powerfulseal/issues/224.

It also should satisfy https://github.com/bloomberg/powerfulseal/issues/222 while making the actions more flexible at the same time.

Resolves https://github.com/bloomberg/powerfulseal/issues/224
Resolves https://github.com/bloomberg/powerfulseal/issues/222
Resolves https://github.com/bloomberg/powerfulseal/issues/241
Resolves https://github.com/bloomberg/powerfulseal/issues/203
Resolves https://github.com/bloomberg/powerfulseal/issues/92

### CODE REVIEW

**Comment by jcstanaway** [comment]
Suggest including #231 in the scope of changes.  Basically, move `filterDayTime` and `filterProbability` to properties of `nodeAction` and `podAction` and are evaluated before `matches`.

File: powerfulseal/policy/action_nodes_pods.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        if not criterion:
+            return []
+        size = criterion.get("size")
+        if size is None:
+            ratio = criterion.get("ratio", 1)
+            size = int(len(candidates)*ratio)
+        if size == 0:
+            self.logger.info("RandomSample size 0")
+            return []
+        return random.sample(candidates, size)
```
Comment (dgzlopes) [inline powerfulseal/policy/action_nodes_pods.py:168]: If the sample size is smaller than candidate's size, a ValueError is raised [[0](https://docs.python.org/3/library/random.html#random.sample)]

Do we check it somewhere?

File: powerfulseal/policy/policy_runner.py [hunk 2]
```diff
@@ -58,42 +59,37 @@ def run(cls, policy, inventory, k8s_inventory, driver, executor,
...
         ]
         while loops is None or loops > 0:
-            for scenario in node_scenarios:
-                scenario.execute()
-            for scenario in pod_scenarios:
-                scenario.execute()
-            sleep_time = int(random.uniform(wait_min, wait_max))
-            logger.info("Sleeping for %s seconds", sleep_time)
-            time.sleep(sleep_time)
+            if len(scenarios) == 0:
```
Comment (dgzlopes) [inline powerfulseal/policy/policy_runner.py:80]: Ultra nit. Using the implicit booleanness of the empty list is more Pythonic [[0](https://docs.python.org/3/library/stdtypes.html#truth-value-testing)]

File: powerfulseal/policy/action_nodes_pods.py (hunk 1 above)
Comment (seeker89) [reply powerfulseal/policy/action_nodes_pods.py:162]: Good find. Thanks!

File: docs/getting-started.md [hunk 3]
```diff
@@ -26,7 +26,7 @@ powerfulseal --help # or seal --help
 
 To start the web interface, use flags `--server --server-host [HOST] --server-port [PORT]` when starting PowerfulSeal in autonomous mode and visit the web server at `http://HOST:PORT/`.
 
-Python 3.6, Python 3.7 and Python 3.8 are supported.
```
Comment (jcstanaway) [inline docs/getting-started.md:29]: Dropping support for 3.6 may be an issue for CentOS users where 3.6 is the latest officially supported version of python. Was there a specific need to drop this support?

File: powerfulseal/policy/ps-schema.yaml [hunk 4]
```diff
@@ -70,90 +82,217 @@ definitions:
...
+        additionalProperties: false
+        properties:
+          target:
+            type: object
+            oneOf:
+            - "$ref": "#/definitions/targetURL"
+            - "$ref": "#/definitions/targetService"
+            - "$ref": "#/definitions/targetIngress"
+          retries:
+            type: number
```
Comment (jcstanaway) [inline powerfulseal/policy/ps-schema.yaml:263]: min value?

File: powerfulseal/policy/ps-schema.yaml [hunk 5]
```diff
@@ -226,28 +368,30 @@ definitions:
         properties:
           size:
             type: number
+            minimum: 1
           ratio:
             type: number
+            minimum: 0
```
Comment (jcstanaway) [inline powerfulseal/policy/ps-schema.yaml:374]: max 1 like `filterProbability` below?

File: powerfulseal/policy/ps-schema.yaml [hunk 6]
```diff
@@ -371,6 +525,8 @@ definitions:
 
   actionWait:
     type: object
+    description: >
+      Wait some seconds.
```
Comment (jcstanaway) [inline powerfulseal/policy/ps-schema.yaml:529]: min/max for `seconds` below?

File: docs/getting-started.md (hunk 3 above)
Comment (seeker89) [reply docs/getting-started.md:29]: To streamline and minimise the cost of ownership. If you can't get python 3.7 you can run the docker image, which is probably more convenient anyway.

We've been running python 3.7 as the base of the docker image for a while now.

File: powerfulseal/policy/ps-schema.yaml (hunk 5 above)
Comment (seeker89) [reply powerfulseal/policy/ps-schema.yaml:379]: Nice catch. Added.

File: powerfulseal/policy/ps-schema.yaml (hunk 6 above)
Comment (seeker89) [reply powerfulseal/policy/ps-schema.yaml:536]: 👍

File: powerfulseal/policy/ps-schema.yaml (hunk 4 above)
Comment (seeker89) [reply powerfulseal/policy/ps-schema.yaml:263]: 👍

File: docs/getting-started.md (hunk 3 above)
Comment (jcstanaway) [reply docs/getting-started.md:29]: In our company, we have a standard, harden CentOS base image we are supposed to be using for all of our images, so using the pre-built seal image isn't an option.  We'll have to install 3.7 (or more likely 3.8) from source rather than a simple `yum install`.  Not a major issue, but something I suppose we'll have to deal with.

File: tests/policy/test_action_kubectl.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+from tests.fixtures import action_kubectl
+
+
+def test_creates_cleanup_action(action_kubectl):
+    action_kubectl.schema["autoDelete"] = True
+    action_kubectl.execute()
+    cleanup = action_kubectl.get_cleanup_actions()
+    assert len(cleanup) == 1
+    job = cleanup[0]
+    print(job)
```
Comment (dgzlopes) [inline tests/policy/test_action_kubectl.py:32]: Nit :)

**Comment by dgzlopes** [comment]
LGTM :rocket:

**Comment by seeker89** [comment]
> Suggest including #231 in the scope of changes. Basically, move `filterDayTime` and `filterProbability` to properties of `nodeAction` and `podAction` and are evaluated before `matches`.

I'll address this one separately.
